### PR TITLE
refactor(source): minor refactors on source parser

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

mostly cleanup
- move json schema to codec crate #17002
- cleanup avro related code in connector crate

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17184?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#17101** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17101?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#17184** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17184?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#17110** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17110?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/connector/codec/src/lib.rs [hunk 1]
```diff
@@ -42,3 +42,16 @@
...
+pub use apache_avro::schema::Schema as AvroSchema;
+pub struct JsonSchema(pub serde_json::Value);
+impl JsonSchema {
+    pub fn parse_str(schema: &str) -> anyhow::Result<Self> {
+        use anyhow::Context;
+
+        let value = serde_json::from_str(schema).context("failed to parse json schema")?;
+        Ok(Self(value))
+    }
+}
```
Comment (xxchan) [inline src/connector/codec/src/lib.rs:57]: exposed these while developing codec-console.

File: src/connector/Cargo.toml [hunk 2]
```diff
@@ -125,7 +119,11 @@ rustls-native-certs = "0.7"
 rustls-pemfile = "2"
 rustls-pki-types = "1"
 rw_futures_util = { workspace = true }
-sea-schema = { version = "0.14", features = ["default", "sqlx-postgres", "sqlx-mysql"] }
+sea-schema = { version = "0.14", default-features = false, features = [
+    "discovery",
+    "sqlx-postgres",
+    "sqlx-mysql",
+] }
```
Comment (xxchan) [inline src/connector/Cargo.toml:126]: This is changed because I found `cargo check -p risingwave_connector` compile failed.

**Comment by BugenZhao** [comment]
> * delete unused `update` op

I would suggest keeping it. There's no problem at all for us to generate an `Update` record for formats like Debezium (and it _was_ before the unified parser refactoring). Compared to the current approach to use `Upsert` everywhere, a retractable record with correct old value can be beneficial in some places I guess.

File: src/connector/codec/src/lib.rs [hunk 3]
```diff
@@ -42,3 +42,16 @@
 /// Converts JSON/AVRO/Protobuf data to RisingWave datum.
 /// The core API is [`decoder::Access`].
 pub mod decoder;
+
+pub use risingwave_pb::plan_common::ColumnDesc;
+pub type RisingWaveSchema = Vec<ColumnDesc>;
```
Comment (BugenZhao) [inline src/connector/codec/src/lib.rs:47]: Does the name sound too general to the outsiders? 😄

File: src/connector/codec/src/decoder/json/mod.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+/// FIXME: when the JSON schema is invalid, it will panic.
+///
+/// ## Notes on type conversion
+/// Map will be used when an object doesn't have `properties` but has `additionalProperties`.
+/// When an object has `properties` and `additionalProperties`, the latter will be ignored.
+/// <https://github.com/mozilla/jsonschema-transpiler/blob/fb715c7147ebd52427e0aea09b2bba2d539850b1/src/jsonschema.rs#L228-L280>
+///
+/// TODO: examine other stuff like `oneOf`, `patternProperties`, etc.
+pub fn json_schema_to_columns(json_schema: &serde_json::Value) -> anyhow::Result<Vec<ColumnDesc>> {
```
Comment (BugenZhao) [inline src/connector/codec/src/decoder/json/mod.rs:28]: Can we make it a method on the new type?

File: src/connector/codec/src/lib.rs (hunk 3 above)
Comment (xxchan) [reply src/connector/codec/src/lib.rs]: removed. Add later if needed

File: src/connector/src/parser/json_parser.rs [hunk 5]
```diff
@@ -93,27 +90,13 @@ pub async fn fetch_json_schema_and_map_to_columns(
...
-/// Map will be used when an object doesn't have `properties` but has `additionalProperties`.
-/// When an object has `properties` and `additionalProperties`, the latter will be ignored.
-/// <https://github.com/mozilla/jsonschema-transpiler/blob/fb715c7147ebd52427e0aea09b2bba2d539850b1/src/jsonschema.rs#L228-L280>
-///
-/// TODO: examine other stuff like `oneOf`, `patternProperties`, etc.
-fn json_schema_to_columns(json_schema: &serde_json::Value) -> ConnectorResult<Vec<ColumnDesc>> {
-    let avro_schema = convert_avro(json_schema, Context::default()).to_string();
-    let schema = Schema::parse_str(&avro_schema).context("failed to parse avro schema")?;
-    avro_schema_to_column_descs(&schema, Some(MapHandling::Jsonb)).map_err(Into::into)
+    json_schema.json_schema_to_columns().map_err(Into::into)
```
Comment (xiangjinwu) [inline src/connector/src/parser/json_parser.rs:99]: nit: `json_schema_` in method name feels redundant.

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**